crates/gazette/src/journal/append.rs

```diff
@@ -0,0 +1,111 @@
+use super::Client;
+use crate::{journal::check_ok, Error};
+use futures::{FutureExt, Stream, StreamExt};
+use proto_gazette::broker::{self, AppendResponse};
+use std::sync::Arc;
+
+impl Client {
+    /// Append the contents of a byte stream to the specified journal.
+    /// Returns a Stream of results which will yield either:
+    /// - An AppendResponse after all data is successfully appended
+    /// - Errors for any failures encountered.
+    /// If polled after an error, regenerates the input stream and
+    /// retries the request from the beginning.
+    pub fn append<'a, S>(
+        &'a self,
+        request: broker::AppendRequest,
+        source: impl Fn() -> S + Send + Sync + 'a,
+    ) -> impl Stream<Item = crate::Result<broker::AppendResponse>> + '_
+    where
+        S: Stream<Item = std::io::Result<bytes::Bytes>> + Send + 'static,
+    {
+        let source = Arc::new(source);
+        coroutines::coroutine(move |mut co| async move {
+            loop {
+                let input_stream = source();
+
+                match self
+                    .append_all(request.clone(), Box::pin(input_stream))
```

Comment:
I believe you can remove the Unpin constraint on append_all, and then this Box::pin